src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,351 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if !NETSTANDARD2_0
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Internal.Log;
+
+namespace Microsoft.CodeAnalysis.Extensions;
+
+[ExportWorkspaceService(typeof(IExtensionMessageHandlerService)), Shared]
+internal sealed class ExtensionMessageHandlerService : IExtensionMessageHandlerService, IDisposable
+{
+    /// <summary>
+    /// Extensions assembly load contexts and loaded handlers, indexed by handler file path. The handlers are indexed by type name.
+    /// </summary>
+    private readonly Dictionary<string, CustomMessageHandlerExtension> _extensions = new();
+
+    /// <summary>
+    /// Handlers of document-related messages, indexed by handler message name.
+    /// </summary>
+    private readonly Dictionary<string, IExtensionDocumentMessageHandlerWrapper> _documentHandlers = new();
```

Comment:
these might be simpler as a multi-dictionary.  though i get your desire to only have one, and error if another extension loads the same name.  I don't mind actually having that happen, us logging a warning about duplicate registration,a nd just deferring to the first extension that registered the name.

Reply:
I really don't want an extension to risk invoking another extension's handler by mistake. I'll leave it like this unless you feel strongly otherwise